### DISCUSSION TITLE
I2S and WiFi interference

### DISCUSSION THREAD

**Discussion by vortigont** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#258] [General]
Hi all,
inspired by #256 (but not necessary related)
Has anyone ever noticed generic WiFi issues with esp32 when I2S engine running at clock speed 10 and specially 20 MHz?
I remember facing this severely sometime ago. With running i2s a simple pinging board's IP becomes spiky with a huge jitter. For quite a while I was even thinking that some of my boards are faulty. Than using [Hallard's](https://github.com/hallard/WeMos-Matrix-Shield-DMA) shield made life a little bit easier, so I forgot about it. Currently I do not have any LED boards at hand, so can't test or confirm it. Just want it to be here to be confirmed or dis-proofed by others.
Maybe tweaking clock speed to something other than multipliers of 10 MHz could make a difference.

Cheers!
cc @mrfaptastic @solhuebner

**Comment by mrcodetastic** [comment]
Interesting. So you're thinking this isn't a SoC DMA / silicon issue causing internal bandwidth contention with the WIFI stack, but actually RF noise generated by GPIO output possibly impacting WiFi performance?

So, potentially something to do with harmonics?

**Reply by board707** [reply]
> So you're thinking this isn't a SoC DMA / silicon issue causing internal bandwidth contention with the WIFI stack, but actually RF noise generated by GPIO output possibly impacting WiFi performance?

I don’t think that this is RF noise, since on the other lib the output to the matrix does not interfere with Wi-Fi.
I just checked that on PxMatrix I can stream lowres video to 64x32 panel via ESP32 WiFi with 25 fps with only minor artifacts

**Comment by vortigont** [comment]
yep, might be that way. First I suspected that something is wrong here when working with my [testbed](https://github.com/vortigont/esp32-hub75-testbed). Periodic UI updates were quite unstable, sometime better in AP mode, sometimes not. Do not remember if it was before we dropped clock rates other than 10/20 MHz or later. But it definitely worth a closer look next time.

**Comment by vortigont** [comment]
Found similar issue, just leave it here for crossreference espressif/arduino-esp32#5834

**Comment by ergindemir** [comment]
What kind of devkit/pcb are you using? Most devkits are poorly designed with signal traces on the bottom of the pcb.

**Comment by vortigont** [comment]
It was Wemos D1 esp32 with and without [Hallard's](https://github.com/hallard/WeMos-Matrix-Shield-DMA) shield.
Boards are OK, since I've used it before for other projects with no similar issues.

**Reply by ergindemir** [reply]
Ideally all signal traces should be on the top with pure ground on the bottom plane. No gpio pins next to the antenna. In addition the board should come with a proper regulator, high output capacitor and decoupling capacitor next to the module. Finally, you should check if your board uses a genuine or knock off wroom module. If you can eliminate all these factors you may suspect that the 12th or 24th harmonic of the i2s signal causes the issue. From my experience devkit v1 module works fine despite knock off module, low output cap, few signal traces on the bottom and few pins next to the antenna. Another devkit with 38 pin which suffers from all causes above has severe wifi issues.
![image](https://user-images.githubusercontent.com/22247668/163168995-3aa17898-de0b-4b9f-a151-5c3fe4d51bb5.png)

**Comment by xavirg** [comment]
I'm facing this same issue with an Adafruit Matrix Portal S3 board.

With the display initialized, wifi is unstable and I get timeouts on my websockets clients or my HTTP client, both fail from time to time and then reconnect.

If I don't init the display, connections run smooth without a single failure.

This has been driving me nuts. I even tried with different power supplies, cables and even changed the AP 😤

**Reply by mrcodetastic** [reply]
Can you try a different board to first make sure it's not something specific to that board? Also try reducing the clock speed to 8Mhz for the HUB75 output.

**Reply by xavirg** [reply]
Isn't the clock speed [already set to 8Mhz by default](https://github.com/mrcodetastic/ESP32-HUB75-MatrixPanel-DMA/blob/master/src/ESP32-HUB75-MatrixPanel-I2S-DMA.h#L324)?

Gonna test my board manually setting the speeds as I'm waiting for another board. I'll come back to you.
```
HUB75_I2S_CFG mxconfig(64, 32, 1, pins);
mxconfig.i2sspeed = HUB75_I2S_CFG::HZ_8M;
```

**Reply by xavirg** [reply]
Seems like setting the clock to 20MHz solves my problem 🚀 . With `HUB75_I2S_CFG::HZ_20M;`, everything works as expected:
```
I (5254) I2S-DMA: Free heap: 278828
I (5264) I2S-DMA: Free SPIRAM: 0
I (5264) I2S-DMA: allocating rowBitStructs with pixel_color_depth_bits of 8
I (5274) I2S-DMA: Allocating 16384 bytes memory for DMA BCM framebuffer(s).
I (5274) I2S-DMA: Minimum visual refresh rate (scan rate from panel top to bottom) requested: 60 Hz
W (5284) I2S-DMA: lsbMsbTransitionBit of 0 gives 76 Hz refresh rate.
I (5294) I2S-DMA: DMA has pixel_color_depth_bits of 8
I (5294) I2S-DMA: Recalculated number of DMA descriptors per row: 128
I (5304) I2S-DMA: 2048 DMA descriptors linked to buffer data.
I (5304) S3: Clock divider is 7
I (5314) I2S-DMA: DMA setup completed
I (1754) main_task: Returned from app_main()
I (8894) websocket: got data (124 bytes)
I (8894) draw: drawing webp (1 frames)
I (8904) draw: finished
I (28764) websocket: got data (3406 bytes)
I (28764) draw: drawing webp (12 frames)
I (29324) draw: finished
I (48964) websocket: got data (20896 bytes)
I (48974) draw: drawing webp (200 frames)
I (68874) draw: finished
I (68904) websocket: got data (124 bytes)
I (68904) draw: drawing webp (1 frames)
I (68914) draw: finished
```


However, if I set it to 8MHz or keep it unsetted I get two more warnings about `lsbMsbTransitionBit`. Then the wifi connection does some negotiations and the websocket client starts to fail:
```
I (1694) I2S-DMA: Free heap: 278824
I (1704) I2S-DMA: Free SPIRAM: 0
I (1704) I2S-DMA: allocating rowBitStructs with pixel_color_depth_bits of 8
I (1714) I2S-DMA: Allocating 16384 bytes memory for DMA BCM framebuffer(s).
I (1714) I2S-DMA: Minimum visual refresh rate (scan rate from panel top to bottom) requested: 60 Hz
W (1724) I2S-DMA: lsbMsbTransitionBit of 0 gives 30 Hz refresh rate.
W (1734) I2S-DMA: lsbMsbTransitionBit of 1 gives 61 Hz refresh rate.
W (1734) I2S-DMA: lsbMsbTransitionBit of 1 used to achieve refresh rate of 60 Hz. Percieved colour depth to the eye may be reduced.
I (1754) I2S-DMA: DMA has pixel_color_depth_bits of 7
I (1754) I2S-DMA: Recalculated number of DMA descriptors per row: 64
I (1764) I2S-DMA: 1024 DMA descriptors linked to buffer data.
I (1764) S3: Clock divider is 16
I (1774) I2S-DMA: DMA setup completed
I (1784) main_task: Returned from app_main()
I (2264) wifi:<ba-add>idx:1 (ifx:0, 70:3a:cb:29:e9:27), tid:0, ssn:1, winSize:64
I (4794) wifi:<ba-del>idx:1, tid:0
I (5294) wifi:<ba-add>idx:1 (ifx:0, 70:3a:cb:29:e9:27), tid:0, ssn:3, winSize:64
I (6324) wifi:<ba-del>idx:1, tid:0
I (6974) wifi:<ba-add>idx:1 (ifx:0, 70:3a:cb:29:e9:27), tid:0, ssn:4, winSize:64
I (14044) wifi:<ba-del>idx:1, tid:0
I (14144) wifi:<ba-add>idx:1 (ifx:0, 70:3a:cb:29:e9:27), tid:0, ssn:7, winSize:64
I (15034) wifi:<ba-del>idx:1, tid:0
I (15064) wifi:<ba-add>idx:1 (ifx:0, 70:3a:cb:29:e9:27), tid:0, ssn:9, winSize:64
E (34634) transport_ws: Error read data
E (34634) transport_ws: Error reading payload data
I (44974) websocket: got data (109 bytes)
E (44974) draw: failed to create WebP decoder
W (44994) transport_ws: esp_transport_ws_poll_connection_closed: unexpected data readable on socket=54
W (44994) websocket_client: Connection terminated while waiting for clean TCP close
```

**Reply by JetForMe** [reply]
I have the same problem with the Adafruit MatrixPortal S3. I've tried 8, 15, and 20 MHz without success.

**Reply by w531t4** [reply]
same

**Reply by mrcodetastic** [reply]
Looks like a common problem with this board. 

Probably the way they've created the PCB there's some resonance or noise.

**Comment by JetForMe** [comment]
Not sure if I should start a new discussion or continue here. But I tried my same code using Protomatter instead of this library, and WiFi stays reliable. I'm not sure why, but on the Adafruit MatrixPortal S3 using this library, I can't get any speed from 1 MHz to 20 MHz to not kill wifi.

**Reply by mrcodetastic** [reply]
Protomatter doesn't use DMA, but if that library works for you then perhaps that's the only option. I'm not sure why the Adafruit MatrixPortal S3  doesn't work with this library... I suspect the PCB they've designed isn't well suited (i.e. generates noise) for high-frequency continuous signal output that this library produces, and as such, kills the WiFi RF chain.

**Reply by JetForMe** [reply]
Sadly with Protomatter I get some glitches in the update. I just have a all green LEDs at 40/255 brightness, no other updates, and you can see very frequent (but brief) flashes on horizontal lines, sometimes the whole panel. Not sure if that would've shown up with DMA. I'm also not sure why it's happening at all, except that perhaps my other busy task is on the same core. I haven't tried pinning it to the other core to to see if I can keep Protomatter on one and this on the other. I'm not actually sure how Protomatter works, except that it must be interrupt-driven as I'm not calling anything in the loop.

**Reply by JetForMe** [reply]
I [posted](https://forums.adafruit.com/viewtopic.php?t=216913) on the Adafruit forums to see if they might be able to offer advice or spin a new, more robust board.

**Reply by mrcodetastic** [reply]
Yep. Glitches or other visual issues will absolutely occur as their library is essentially an endless loop using the CPU to bitbang the output. So any other activities the CPU needs to do, even got a microsecond will interrupt this and it'll be visible by the eye.

**Reply by JetForMe** [reply]
I did some experimenting with pinning things to cores, but it had no effect. I narrowed it down to simply putting data onto a FreeRTOS queue in the ISR, not even reading it in the main loop(), that was enough to cause the glitch.

**Reply by JetForMe** [reply]
It's still curious that their approach doesn't kill wifi, but this one does.

**Reply by xavirg** [reply]
> It's still curious that their approach doesn't kill wifi, but this one does.

Did you try combining all the available led drivers with all clock speeds?

That's how I found I needed to use FM6124 driver with 20MHz to make it work. Tried all speeds with all drivers.

**Reply by JetForMe** [reply]
Oh huh it didn't occur to me to try a different driver, I missed that. I can't remember if this is an Adafruit panel or one I got off Amazon, but hopefully a different driver works, and doesn't cause interference.

**Reply by JetForMe** [reply]
I tried them all, sadly, nothing worked. At first, `FM6124` seemed to work, but then it faltered. I went through all of them and saw no real improvement. I went back to `FM6124`, and it was as bad as ever.

**Reply by JetForMe** [reply]
@mrcodetastic I was looking through the Adafruit Protomatter code to see how it operates, and it uses DMA! Not sure how it differs from yours, but it’s definitely using DMA for something. https://github.com/adafruit/Adafruit_Protomatter/blob/d55e3bf83306b19651c1905bfb984f3768c2569e/src/arch/esp32-s3.h#L388

**Reply by mrcodetastic** [reply]
Haha. I stand corrected. In fact, if I remember correctly, I actually use this code in this library for the s3 implementation as well.

Clearly, it was a long time ago I implemented it.

Depends if they just DMA for an endless loop, or have an isr constantly triggering a DMA output. I haven't looked at their code in detail.

**Reply by JetForMe** [reply]
I didn't look in detail either, only enough to see some DMA setup, and a timer set up. But theirs is definitely glitchy, and worse: I set up code in the main Arduino loop to just fill the display (at a brightness of 40) with a series of colors, changing every second, and now the MatrixPortal S3 makes an annoying whine which varies based on which color is displayed (I couldn't perceive it when displaying only pure green). Yours doesn't make this whine. With the same basic setup, it does seem a bit dimmer. And of course, it kills WiFi. I wonder why theirs doesn't. Am I gonna have to take the time to dig into both of these? lol

**Reply by JetForMe** [reply]
I suppose another approach would be to look at the waveforms on the 'scope and see how they differ. Are we sure this is an EMI issue, or is it something in the software stack that’s disrupting WiFi?

**Reply by mrcodetastic** [reply]
If you use a scope it will take about one minute to figure out whether protonmatter is using DMA or not. If you see a break in the clock signal then it's not DMA... Because it's being bit banged and subject to interruption.

**Reply by JetForMe** [reply]
(I'm doing some investigating of what's happening, lemme know if I should start a new discussion for this).

How does your library work on S3 at a high level? Does it set up timers and interrupts to trigger DMA? Are these things pinned to a core? What’s their priority? Is it possible there’s contention with the WiFi and TCP tasks (which are pinned to core 0)?

**Reply by JetForMe** [reply]
There are definitely breaks in the CLK signal using Protomatter. They're operating at about 17 MHz, and there’s a *lot* of jitter between bursts, but it’s typically 30 µs between them.

**Reply by mrcodetastic** [reply]
> How does your library work on S3 at a high level? 

DMA

> Does it set up timers and interrupts to trigger DMA? Are these things pinned to a core? What’s their priority? Is it possible there’s contention with the WiFi and TCP tasks (which are pinned to core 0)?

None of this. It's DMA a continuous stream (loops continuously) of binary out. No CPU involvement, no interrupts etc.

**Reply by mrcodetastic** [reply]
I suggest reducing the color depth of this library to say 3 bits. See if that does anything

**Reply by JetForMe** [reply]
Okay, after days, and many hours of experimenting, I'm fairly confident that this is an issue with the WROOM-1 module, and not likely to be any kind of processor resource contention issue. I've got it working fairly reliably with UM Pro S3 and Feather S3 modules, with both on-board chip antenna and U.FL antenna.

When I first started trying the other boards, I was still experiencing the problems. Thinking it must be resource contention, I wrote my own Web Sockets client to sit atop AsyncTCP, hoping that would improve the CPU core utilization. Things started to work, so I tried it on the WROOM-1 board again, and it failed as it always does. I went back to the UM boards, and it continues to work.

I'm unable to determine if there’s still some interference, but it seems to be performing adequately for my needs, for now. I have yet to see if it will work in the final deployment environment, but I have no real reason to believe it won’t.

**Reply by xavirg** [reply]
> I've got it working fairly reliably with UM Pro S3 and Feather S3 modules, with both on-board chip antenna and U.FL antenna.
> 

Which CPU do they use? I also see they both have 8MB of extra PSRAM.

The MatrixPortal S3 only has 2MB of memory, which results in having about 200kB available at runtime.

**Reply by victorhooi** [reply]
I'm also trying to do some experiments with HUB75 panels ATM, and trying to pick the right controller.

The [Adafruit Matrix Portal S3](https://www.adafruit.com/product/5778) seemed like a good option - however, then I read here that there might be wifi RF/EMI issues with it - and it seems like the SRAM (2MB) amount is too low?

Is the memory amount not enough? And did anybody definitive prove it was a wifi EMI issue? (I think I have a SDR around somewhere - and I'm happy to order a Adafruit Matrix Portal S3, and do some tests, if that would help at all? Definitely not an RF or embeeded expert/guru though - but happy to try to do any tests that people think might help here).

Apollo Automation also has their new [M-1 LED panel and controller](https://apolloautomation.com/products/m-1-led-matrix) - which also seems to be an ESP32-S3 - not sure if that would have similar issues?

Or are there alternative HUB75 controllers, or ESP32-to-HUB75 breakout boards that people recommend?

**Reply by mrcodetastic** [reply]
Do you only want to drive a LED panel? If so, might be worth looking at using a WF2: https://github.com/mrcodetastic/HD-WF1-WF2-LED-MatrixPanel-DMA

Note: If you use one of these devices you won't get access to any of the other GPIOs given it's a board designed only for driving an LED panel.

**Reply by xavirg** [reply]
> The [Adafruit Matrix Portal S3](https://www.adafruit.com/product/5778) seemed like a good option - however, then I read here that there might be wifi RF/EMI issues with it - and it seems like the SRAM (2MB) amount is too low?
> 
> Is the memory amount not enough?


Memory is sufficient. I had the wifi issue just starting my program by initializing wifi and then initializing the display, nothing else.
